### PR TITLE
Team Puffy Shirts bring you (parens)

### PR DESCRIPTION
**Pull request by miriam-cortes**
From the people who brought you the "Charles in Charge" meme, we bring you http://notyetsy.herokuapp.com/

### CODE REVIEW

**Comment by miriam-cortes** [comment]
What helped us stay organized and working together?
- I think the fact that we spoke for a while about our working styles before starting the actual work really helped team Puffy Shirts! Even through tough times we had a "work hard, play harder" attitude that helped us push through the tougher moments. I totally loved working with these people, even when they needed time locked away (from me) in a dark room :)

What should you have been doing differently?
- More tests? Started working on design upgrades? IDK, really. 

What would you recommend to C7? 
- Take time away from your teammates when you need it so you can reconvene in a more refreshed state of mind. Talk through your merges so everyone in the team knows what was changed- it makes finding things that will break much easier.

What went well?
- Everyone pulled their weight; everyone was very proactive and always asked what else they could start working on. It really made the Task Master job easier.

What was most challenging?
- The whole order/cart situation. We didn't realize what a HUUUUUUUGE part that was so when Sabrina took it on (by herself) we thought nothing of it. As time passed we finished our individual portions and wanted to help out with hers but it would've taken so much extra time catching us up that there was really nothing we could do immediately. 

How do you feel about your project overall?
- We rock. I'm glad we were able to end this project on a high note! Seriously, C6 is pretty darn awesome!

**Comment by wilsab30** [comment]
**What helped you stay organized and working together?**
We had many check-ins and had a routine of pushing, pull requests, reviewing and merging. This helped us all stay on track and knowing what the next step would be.

**What should you have been doing differently?**
I don't know that as a group we should have done anything differently. We disagreed, but also laughed, sang and worked it out.

**One piece of advice for c7.** 
Take the time to talk it all out before writing any code . We did and I think that's why we had fun and got stuff done.

**What went well technically?**
We didn't have large merge conflicts. We keep things pretty contained and were able to workout any problems.

**What was most challenging?**
The #$%## cart. We had no idea how different that would be from the apps we had built before. I felt so bad about not having more done sooner. I was complicated in so many ways. I wish we had some warning before starting. Maybe we would have broken that part up among a few people.

Overall I think we did a mashing job of it. Not only did we make a really cute site, that works, we also laughed together and voiced appreciation for each other along the way.

**Comment by ninamutty** [comment]
Team Work:
- I found the stand-ups helpful in staying organized and I think we did a good job of communicating where each us were in our work - both what we had completed and what we needed to accomplish. I think that we did a good job trouble shooting together as well. 
- The first week was a little tougher both in stress levels and communication. I was out sick the first Monday (and as a result didn't understand sessions or the cart) so I had a hard time helping with problems until I was able to catch back up. I think we also didn't always know the best way to help each other in the beginning, but that really changed in the second week and the team dynamic felt much more cohesive.
- Have two people working on the cart. It seemed like a much bigger part than any of us were anticipating, particularly because the other aspects of the project were more familiar. I think it would have been helpful for two people to understand and work through it (especially in the beginning) and the other two dividing up some of the more familiar problems.

Technical:
- I primarily worked on the products and reviews pages and the styling (except for the merchant and cart pages). I felt good about the functionality I added for the products and reviews and how that changed based on if the user was a merchant or not. I also got more familiar with Foundation which I was happy about.
- I found the testing challenging and I know I didn't test for all the edge cases. I think I handled the possibilities decently well in the model and controller but didn't test for them which I really wish I had time to do. I also which some aspects the site were a little more responsive, but overall I'm really happy with it!

Overall:
I really enjoyed this project! While we had our low points, I think my team worked well together and I had a lot of fun building NotYetsy. I'm definitely proud of it and the work we did together!

**Comment by livmaria7891** [comment]
How did your team work together:
1. What helped you stay organized and working together?
  Our team used a trello board to keep an organized 'to do list' and a spread sheet to divide tasks into waves. Every day we reviewed code changes and pull requests as group to make sure there weren't any serious conflicts. We also communicated a lot.
1. What should you have been doing differently
   Even though we utilized a trello board, it got a little disorganized at times. I think we also could have split work differently because some people carried a far heavier load than others. 
2. What would you recommend to C7? Once piece of advice.
   Our team made moral a priority. Every time we became frustrated or discouraged our team had silly rituals in place that helped to make us happy. By the end of the project we had grown closer together and enjoyed each others company more rather than being tired of each other and I think that was because we made team happiness our top priority. 

Technical 
1. What went well? 
  We divided the work in such a way that merge conflicts were rare. It really helped to keep our product working most of the time, even when we were merging lots of code. 
1. What was the most challenging?
   The most challenging piece was watching Sabrina suffer through her part of the project and none of us having the knowledge to help. We also ran out of time to write thorough tests.  

How do you feel about your project overall?
 I am very pleased with our project and how well my team worked together. Although our finished product isn't perfect it does just about everything we wanted it to and it looks great.

File: test/models/category_test.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+require 'test_helper'
+
+class CategoryTest < ActiveSupport::TestCase
+  # test "the truth" do
+  #   assert true
+  # end
+
```
Comment (CheezItMan) [inline test/models/category_test.rb:7]: You should also test if the name is set correctly it's valid.

File: test/models/merchant_test.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
+require 'test_helper'
+
```
Comment (CheezItMan) [inline test/models/merchant_test.rb:2]: You might also include tests for merchants who don't have any orders, or products, etc.

File: test/models/product_test.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    assert_not products(:glitter).valid? "Glitter should no longer be valid"
+
+    products(:new_shoe).price = nil
+    assert_not products(:new_shoe).valid? "New Shoe should no longer be valid"
+  end
+
+  test "Products cannot have a rating less than 0 or greater than 5" do
+    assert products(:glitter).rating = 6
+    assert_not products(:glitter).valid? "Glitter should no longer be valid"
+
```
Comment (CheezItMan) [inline test/models/product_test.rb:21]: How does this work, isn't rating a method?

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -2,4 +2,49 @@ class ApplicationController < ActionController::Base
...
+    @all_categories = Category.all
+    @lesser_categories = find_lesser_categories(ordered_categories, limit)
+  end
+
+  def find_lesser_categories(ordered_categories, limit)
+    lesser_categories = ordered_categories[limit..-1]
+    return lesser_categories
+  end
+
+  def current_cart
```
Comment (CheezItMan) [inline app/controllers/application_controller.rb:32]: It's interesting that you have a different cart for merchant or guest.  That will work and it lets you merge carts when they log in finally.  Tricky.

File: app/controllers/categories_controller.rb [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  def index
+    # @category = Category.all
+    @category = Category.order("lower(name) ASC").all
+    @new_category = Category.new
+
+
+  end
+
+  def show
+    @category = Category.find(params[:id].to_i)
```
Comment (CheezItMan) [inline app/controllers/categories_controller.rb:11]: What about when you can't find the Category, it's been deleted etc?

File: app/controllers/products_controller.rb [hunk 6]
```diff
@@ -0,0 +1,90 @@
+class ProductsController < ApplicationController
+  before_action :find_product, only: [:show, :edit, :update]
```
Comment (CheezItMan) [inline app/controllers/products_controller.rb:2]: Good use of filters here.  You could probably DRY it up a bit further, but not much.

File: app/helpers/application_helper.rb [hunk 7]
```diff
@@ -1,2 +1,22 @@
 module ApplicationHelper
+
+
```
Comment (CheezItMan) [inline app/helpers/application_helper.rb:3]: Very good choice in a view helper!

File: app/helpers/orders_helper.rb [hunk 8]
```diff
@@ -0,0 +1,59 @@
+module OrdersHelper
+
```
Comment (CheezItMan) [inline app/helpers/orders_helper.rb:2]: Another good use of a view helper

File: app/models/merchant.rb [hunk 9]
```diff
@@ -0,0 +1,134 @@
+# Merchant Model
```
Comment (CheezItMan) [inline app/models/merchant.rb:1]: This is very very well done.  Good use of validations & business logic here!

File: test/controllers/categories_controller_test.rb [hunk 10]
```diff
@@ -0,0 +1,44 @@
+require 'test_helper'
```
Comment (CheezItMan) [inline test/controllers/categories_controller_test.rb:1]: On your controller tests you should also include some negative cases, what if the user tries to see a category that's been deleted or never existed.  What if someone without permission tries to edit a category?

File: test/controllers/sessions_controller_test.rb [hunk 11]
```diff
@@ -0,0 +1,41 @@
+require 'test_helper'
```
Comment (CheezItMan) [inline test/controllers/sessions_controller_test.rb:1]: You probably also want to test the response if a user goes directly to the callback url without going through github.

File: app/views/merchants/show.html.erb [hunk 12]
```diff
@@ -0,0 +1,61 @@
+<div class="small-12 columns buttons_container row">
```
Comment (CheezItMan) [inline app/views/merchants/show.html.erb:1]: Good work on responsiveness.  You could DRY up your views a bit by using partials, but well done!

**Review by CheezItMan — COMMENTED** [review]
@livmaria7891 @wilsab30 @ninamutty @miriam-cortes 

Hello Puffy Shirts!  I've put in comments & a summary in your weekly progress reports.  Overall well done.  I particularly liked the Cart, Model Validations & custom methods and the styling.  

Things to work on:
- Tests (check negative cases especially)
- Drying up Views with View Helpers & Partials
- Working on access control (preventing users from getting to things they shouldn't.
